src/data/nat/totient.lean

```diff
@@ -335,4 +338,142 @@ begin
+begin
+  apply strong_sub_induction,
+  intros n m hxy,
+  by_cases hco : coprime n m,
+  rw coprime_iff_gcd_eq_one.1 hco,
+  simp only [totient_one, one_mul, mul_one, totient_mul hco],
+  by_cases g1 : n ≠ 0,
+  by_cases g2 : m ≠ 0,
+  obtain ⟨q, hq1, hq2⟩ := prime.not_coprime_iff_dvd.1 hco,
+  haveI : fact q.prime, exact { out := hq1 },
+  have hn1 := padic_val_int_dvd q n,
+  have hm1 := padic_val_int_dvd q m,
+  norm_cast at *,
+  simp only [padic_val_int.of_nat] at *,
+  have hn2 := p_val_nat_div_coprime n q g1 hq1,
+  have hm2 := p_val_nat_div_coprime m q g2 hq1,
+  have hnm2 := coprime.pow_left (padic_val_nat q n + padic_val_nat q m)
+    (p_val_nat_div_coprime (n * m) q (mul_ne_zero g1 g2) hq1),
+  rw padic_val_nat.mul q g1 g2 at hnm2,
+  rw [ (nat.mul_div_cancel' hn1).symm , (nat.mul_div_cancel' hm1).symm],
+  let r := padic_val_nat q n,
+  let s := padic_val_nat q m,
+  have : q ^ r * (n / q ^ r) * (q ^ s * (m / q ^ s)) = q^(r+s) * (n * m / q^(r+s)),
+  by { rw [ (nat.mul_div_cancel' hn1), (nat.mul_div_cancel' hm1)],
+    apply symm,
+    apply nat.mul_div_cancel',
+    convert mul_dvd_mul hn1 hm1,
+    apply pow_add, },
+  simp_rw [this, totient_mul hnm2, totient_mul (coprime.pow_left r hn2),
+   totient_mul (coprime.pow_left s hm2), pow_add, (nat.div_mul_div_comm hn1 hm1).symm,
+    coprime.gcd_mul _ (coprime.pow_left s hm2), nat.gcd_comm,
+    coprime.gcd_mul _ (coprime.pow_left r hn2)],
+  have h33 := (coprime_iff_gcd_eq_one.1 (coprime.pow_left r hm2)),
+  rw nat.gcd_comm at h33,
+  rw h33,
+  rw coprime_iff_gcd_eq_one.1 (coprime.pow_left s hn2),
+  simp_rw [mul_one, one_mul, (gcd_self_pow q s r), totient_mul (coprime.gcd_right (m/q^s)
+    ((coprime.pow_left (min s r) hn2)))],
+  have i1 : n/q^r < n, by { apply (nat.div_lt_self (nat.pos_of_ne_zero g1)),
+    rw ←(one_pow (0 : ℕ)),
+    have hr : 0 < r, by { have := (one_le_padic_val_nat_of_dvd (nat.pos_of_ne_zero g1) hq2.1),
+      linarith},
+    apply pow_lt_pow_of_lt_right (prime.one_lt hq1) hr },
+  have i2 : m/q^s < m, by { apply (nat.div_lt_self (nat.pos_of_ne_zero g2)),
+    rw ←(one_pow (0 : ℕ)),
+    have hs : 0 < s, by { have := (one_le_padic_val_nat_of_dvd (nat.pos_of_ne_zero g2) hq2.2),
+      linarith},
+    apply pow_lt_pow_of_lt_right (prime.one_lt hq1) hs, },
+  rw ←(gcd_self_pow q s r),
+  have V := congr (congr_arg has_mul.mul (hxy (n/q^r) (m/q^s) i1 i2))
+    (totient_pow_mul_self r s hq1),
+  rw [ ←mul_assoc],
+  have st1 : ∀ (a b c d : ℕ), a * b * c * d = b * d * a * c, by { intros a b c d, ring },
+  have st2 : ∀ (a b c d e f : ℕ), a * b * c * d * e * f = b * d * f * a * c * e,
+    by { intros a b c d e f, ring },
+  simp_rw [← mul_assoc] at *,
+  have st3 : (q^r).gcd (q^s) = (q^s).gcd (q^r), by { apply nat.gcd_comm },
+  rw [st1, st2, nat.gcd_comm, ←st3],
+  exact V,
+  { simp only [not_not.1 g2, mul_zero, totient_zero, zero_mul] },
+  { simp only [not_not.1 g1, zero_mul, totient_zero, mul_zero] },
+end
```

Comment:
```suggestion
lemma gcd_prod_double_counting {β : Type*} [comm_monoid β] (m n : ℕ) (f : ℕ → β) :
    (∏ (x : ℕ) in (m.gcd n).factors.to_finset, f x) * ∏ (x : ℕ) in (m * n).factors.to_finset, f x = (∏ (x : ℕ) in m.factors.to_finset, f x) * ∏ (x : ℕ) in n.factors.to_finset, f x :=
begin
  rcases eq_or_ne n 0 with rfl | hm0, { simp },
  rcases eq_or_ne m 0 with rfl | hn0, { simp },
  rw [←@prod_union_inter _ _ m.factors.to_finset n.factors.to_finset, mul_comm],
  congr,
  { apply factors_mul_to_finset; assumption },
  { simp only [←support_factorization, factorization_gcd hn0 hm0, finsupp.support_inf] },
end

lemma totient_mul_gen : ∀ a b: ℕ, φ (a.gcd b) * φ (a * b) = φ a * φ b * (a.gcd b) :=
begin
  intros m n,
  simp only [totient_eq_div_factors_mul],

  have shuffle : ∀ a1 a2 b1 b2 c1 c2 : ℕ, b1 ∣ a1 → b2 ∣ a2 →
    (a1/b1 * c1) * (a2/b2 * c2) = (a1*a2)/(b1*b2) * (c1*c2),
  { intros a1 a2 b1 b2 c1 c2 h1 h2,
    calc
      (a1/b1 * c1) * (a2/b2 * c2) = ((a1/b1) * (a2/b2)) * (c1*c2) : by apply mul_mul_mul_comm
      ... = (a1*a2)/(b1*b2) * (c1*c2) : by { congr' 1, exact div_mul_div_comm h1 h2 } },

  rw shuffle _ _,
  rw shuffle _ _,
  rotate, repeat { apply prod_prime_factors_dvd },
  simp only [gcd_prod_double_counting],
  rw [eq_comm, mul_comm, ←mul_assoc, ←nat.mul_div_assoc],
  exact mul_dvd_mul (prod_prime_factors_dvd m) (prod_prime_factors_dvd n)
end
```

Reply:
I'm sorry to step on your PR, but I wanted to check that the suggestion I made yesterday was reasonable and I ended up finishing a shorter proof of your `totient_mul_gen`.  I'm not suggesting that my code above is ready to PR, I just wanted to show you the shape of the proof.  In particular, I don't know whether my `gcd_prod_double_counting` is worth having as a separate lemma, or if it's better to inline this in `totient_mul_gen` with a `have` statement.

Reply:
Sorry I'm just seeing this now. I think this is probably better than the proof I'd done (I hadn't seen Apostol's proof and just went with the first thing that came to mind and sounded lean-reasonable). If you want to PR your proof then that is fine by me and I can just kill this PR.

Reply:
> If you want to PR your proof then that is fine by me and I can just kill this PR.

Ok, thanks.  I've PRd it as #14842, where I've called the lemma `totient_mul_general`.  If you make this PR depend on #14842 then you should be able to continue to use the lemma before that PR gets merged.